# **Topic:** Data Cleaning

## **Learning Objectives:**

By the end of this lesson, students should be able to:

1. Understand the concept of data cleaning.
2. Identify common data quality issues.
3. Apply basic data cleaning techniques using spreadsheet tools or Python (e.g. pandas).
4. Explain the importance of clean data in analysis and decision-making.


### **1. Introduction (10 minutes)**

#### **What is Data Cleaning?**

Data cleaning (also called data cleansing or scrubbing) is the process of identifying and correcting (or removing) errors and inconsistencies in data to improve its quality and usability.

#### **Why is Data Cleaning Important?**

* Dirty data leads to wrong analysis.
* Machine learning models trained on bad data give poor predictions.
* It improves data accuracy, consistency, and reliability.

#### **2. Common Data Quality Issues (10 minutes)**

| **Issue**                   | **Description**                        | **Example**                         |
| --------------------------- | -------------------------------------- | ----------------------------------- |
| **Missing Values**          | Empty or null values in a dataset.     | Name: John, Age: —                  |
| **Duplicate Records**       | The same entry appears more than once. | Two identical rows.                 |
| **Inconsistent Formatting** | Data is not standardized.              | "Nigeria", "nigeria", "NIGERIA"     |
| **Incorrect Data Types**    | Data stored in the wrong format.       | Age as text: "twenty" instead of 20 |
| **Outliers**                | Unusual or extreme values.             | Income: \$10,000,000                |
| **Typographical Errors**    | Spelling mistakes or wrong entries.    | “Gogle” instead of “Google”         |


#### **3. Techniques for Data Cleaning (20 minutes)**

##### **A. Handling Missing Data**

* Remove rows with missing values.
* Fill missing values with:

  * Mean / Median (for numerical data)
  * Mode / Placeholder (for categorical data)

##### **B. Removing Duplicates**

* Use spreadsheet tools: *Remove duplicates* in Excel/Google Sheets.
* Use Python:

```python
df.drop_duplicates(inplace=True)
```

##### **C. Fixing Inconsistent Formatting**

* Standardize case (upper/lower title).
* Trim white spaces.

Example:

```python
df['country'] = df['country'].str.strip().str.title()
```

##### **D. Converting Data Types**

```python
df['age'] = pd.to_numeric(df['age'], errors='coerce')
```

##### **E. Handling Outliers**

* Use statistical methods (e.g. IQR or Z-score) to detect and possibly remove.
* Visual tools: Boxplot.


#### **4. Practical Example (10 minutes)**

Let’s say we have the following data:

| Name | Age | Country | Income |
| ---- | --- | ------- | ------ |
| John | 25  | Nigeria | 5000   |
| Jane |     | nigeria | 5200   |
| John | 25  | Nigeria | 5000   |
| Mike | 130 | U.S.A   | 10000  |
| Anna | 22  | NIGERIA | —      |

#### Cleaning Steps:

* Fill Jane’s missing age.
* Standardize country names to "Nigeria".
* Remove duplicate row.
* Remove or replace Mike’s age (outlier).
* Handle missing income for Anna.


In [1]:
import pandas as pd
import numpy as np

# Sample data
data = {
    'Name': ['John', 'Jane', 'John', 'Mike', 'Anna'],
    'Age': [25, None, 25, 130, 22],
    'Country': ['Nigeria', 'nigeria', 'Nigeria', 'U.S.A', 'NIGERIA'],
    'Income': [5000, 5200, 5000, 10000, None]
}

df = pd.DataFrame(data)
print("Original Data:")
print(df)



# Cleaning Steps:

# Step 1: Fill Jane’s missing age

# Assume we fill it with the average age (excluding the outlier 130).


# Exclude Mike's age (130) as outlier for mean calculation
filtered_ages = df['Age'][df['Age'] < 100]  # Only normal ages
mean_age = filtered_ages.mean()
df['Age'] = df['Age'].fillna(mean_age)

# Step 2: Standardize country names to “Nigeria”

df['Country'] = df['Country'].str.strip().str.title()

# Unify any known variations
df['Country'] = df['Country'].replace({
    'Nigeria': 'Nigeria',
    'U.S.A': 'USA'
})

# Step 3: Remove duplicate rows

df.drop_duplicates(inplace=True)

# Step 4: Remove or replace Mike’s age (130)

# We'll remove it as it's an unrealistic outlier.

df = df[df['Age'] <= 100]

# Step 5: Handle missing income for Anna

# We’ll fill it with the mean income.

mean_income = df['Income'].mean()
df['Income'] = df['Income'].fillna(mean_income)

# Cleaned Data Output:
print("\nCleaned Data:")
df.reset_index(drop=True, inplace=True)
print(df)



Original Data:
   Name    Age  Country   Income
0  John   25.0  Nigeria   5000.0
1  Jane    NaN  nigeria   5200.0
2  John   25.0  Nigeria   5000.0
3  Mike  130.0    U.S.A  10000.0
4  Anna   22.0  NIGERIA      NaN

Cleaned Data:
   Name   Age  Country  Income
0  John  25.0  Nigeria  5000.0
1  Jane  24.0  Nigeria  5200.0
2  Anna  22.0  Nigeria  5100.0




### **5. Assessment / Class Work (5 minutes)**

#### **Instructions:**

You are given the following sample dataset:

```plaintext
| Name     | Age  | Country     | Income  |
|----------|------|-------------|---------|
| Peter    | 30   | Nigeria     | 4500    |
| Mercy    |      | nigeria     | 4800    |
| Ayo      | 28   | Ghana       |         |
| Peter    | 30   | Nigeria     | 4500    |
| Chioma   | 27   | NIGERIA     | 4700    |
| Ibrahim  | 102  | Nigeria     | 4900    |
```

#### **Task 1: Identify Errors**

From the data above, identify **any two issues**.
Examples of issues could include:

* Missing values
* Duplicates
* Inconsistent formatting
* Outliers

Write your answer in this format:

> **Error 1:**
> Description of the error
>
> **Error 2:**
> Description of the error

---

#### **Task 2: Suggest Fixes**

For each error identified above, suggest how you would fix it.

Continue the format like this:

> **Fix for Error 1:**
> Description of how you would fix the issue (e.g., fill with mean, drop duplicates, convert format)

> **Fix for Error 2:**
> Description of how you would fix the issue

---

#### Example Answer (for reference only):

> **Error 1:** Mercy's age is missing.
> **Error 2:** Country values have inconsistent capitalization (e.g., "nigeria", "NIGERIA").

> **Fix for Error 1:** Fill the missing age with the average of other ages.
> **Fix for Error 2:** Standardize all country names using `.str.title()` or `.str.upper()` in pandas.

In [2]:
import pandas as pd
import numpy as np
import random

# Set random seed for reproducibility
random.seed(42)
np.random.seed(42)

# Sample pool
names = ['Alice', 'Bob', 'Charlie', 'Diana', 'Eve', 'Frank', 'Grace', 'Hannah', 'Ivy', 'Jack']
diagnoses = ['Healthy', 'Diabetes', 'Hypertension', 'Asthma', 'Obesity']

# Generate base clean data
data = {
    'Name': np.random.choice(names, 1900),
    'Age': np.random.randint(18, 90, 1900),
    'BloodPressure': np.random.randint(90, 180, 1900),
    'Diagnosis': np.random.choice(diagnoses, 1900)
}
df = pd.DataFrame(data)

# 1. Introduce Missing Values (10% BloodPressure missing)
num_missing = 200
missing_indices = np.random.choice(df.index, size=num_missing, replace=False)
df.loc[missing_indices, 'BloodPressure'] = np.nan

# 2. Introduce Duplicate Records (5%)
duplicates = df.sample(frac=0.05, random_state=42)
df = pd.concat([df, duplicates], ignore_index=True)

# 3. Introduce Inconsistent Formatting (case variations)
def random_case(text):
    options = [text.lower(), text.upper(), text.capitalize()]
    return random.choice(options)

df['Diagnosis'] = df['Diagnosis'].apply(random_case)
df['Name'] = df['Name'].apply(random_case)

# 4. Incorrect Data Types: Introduce some Ages as text (e.g. 'twenty')
df['Age'] = df['Age'].astype(object)
age_error_indices = np.random.choice(df.index, size=20, replace=False)
for idx in age_error_indices:
    # randomly pick a number word instead of integer
    num_word = random.choice(['twenty', 'thirty', 'forty', 'fifty', 'sixty'])
    df.at[idx, 'Age'] = num_word

# 5. Outliers in BloodPressure (set some extreme values)
outlier_indices = np.random.choice(df.index, size=10, replace=False)
df.loc[outlier_indices, 'BloodPressure'] = np.random.choice([30, 300, 400], size=10)

# 6. Typographical Errors in 'Name' (introduce some misspellings)
def typo(name):
    # Simple typo: duplicate a letter or swap two letters
    if len(name) < 3:
        return name
    pos = random.randint(0, len(name)-2)
    if random.random() < 0.5:
        # duplicate letter
        return name[:pos] + name[pos] + name[pos] + name[pos+1:]
    else:
        # swap letters
        return name[:pos] + name[pos+1] + name[pos] + name[pos+2:]

typo_indices = np.random.choice(df.index, size=30, replace=False)
for idx in typo_indices:
    original_name = df.at[idx, 'Name']
    df.at[idx, 'Name'] = typo(original_name)

# Shuffle rows again
df = df.sample(frac=1, random_state=42).reset_index(drop=True)

# Save to CSV
# df.to_csv("patients.csv", index=False)


### **Spotting data uncleanliness is just as important as fixing it**, and beginners often miss this step.

Here’s a **structured approach you can follow** to **spot unclean data** systematically:

#### ✅ 1. **Load the Dataset**

```python
import pandas as pd

df = pd.read_csv("patients.csv")
```

In [3]:
import pandas as pd

df = pd.read_csv("../../datasets/patients.csv")

#### 🔍 2. **Start with Basic Exploration**

Use `.info()`, `.describe()`, `.head()`, `.tail()`, `.sample()`

```python
print(df.info())       # Shows column types and missing values
print(df.describe())   # Stats for numeric columns
print(df.head())       # Preview first few rows
```

✅ **What to Look For**:

* `Non-null count` < total rows → **Missing Values**
* Weird data types (e.g., object for numbers)
* Large difference between min and max → **Outliers**

In [4]:
print(df.info())       # Shows column types and missing values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1995 entries, 0 to 1994
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Name           1995 non-null   object 
 1   Age            1995 non-null   object 
 2   BloodPressure  1782 non-null   float64
 3   Diagnosis      1995 non-null   object 
dtypes: float64(1), object(3)
memory usage: 62.5+ KB
None


Age shouldnt be an object type. It should an integer type. 
The fact that it's `object` means it likely contains **non-numeric values**, such as:

* Missing values stored as strings like `"unknown"`, `"N/A"`, or empty strings `""`
* Inconsistent formats, like `"30 years"` instead of `30`
* Words instead of numbers, like `"forty"`

### ✅ **How to Fix It**

Here’s a step-by-step approach in code:

In [5]:
# Check unique values to identify bad entries
print(df['Age'].unique())

['23' '49' '19' '74' '85' '58' '81' '84' '69' '77' '30' '72' '88' '18'
 '38' 'thirty' '89' '42' '87' '73' '50' '61' '32' '70' '59' '36' '51' '82'
 '78' '47' '66' '24' '45' '76' 'twenty' '46' '34' '75' '41' '35' '27' '65'
 '80' '20' '55' '21' '37' '43' '53' '57' '29' '25' '44' '64' '67' '48'
 '63' '62' '31' '68' '71' '54' '79' '39' '40' '60' '56' '28' '26' '83'
 '33' '52' '86' '22' 'fifty' 'sixty' 'forty']


As suspected, the `Age` column is a mix of **numeric strings** (e.g., `'23'`, `'49'`) and **word-based numbers** like:

* `'thirty'`
* `'twenty'`
* `'forty'`
* `'fifty'`
* `'sixty'`

These word-based entries are the reason the column is stored as `object` instead of a numeric type.

---

### ✅ Here's How to Clean and Convert `Age` Properly

You can map the word-based ages to their numeric equivalents and then convert the column to `float` or `int`.

In [6]:
# Mapping for word-based numbers
word_to_num = {
    'twenty': 20,
    'thirty': 30,
    'forty': 40,
    'fifty': 50,
    'sixty': 60
}

# Replace word-based entries with corresponding numeric values
df['Age'] = df['Age'].replace(word_to_num)

# Now convert the entire column to numeric (this will coerce any remaining bad values to NaN)
df['Age'] = pd.to_numeric(df['Age'], errors='coerce')

# Confirm conversion
print(df.dtypes)
print(df['Age'].unique())

Name              object
Age                int64
BloodPressure    float64
Diagnosis         object
dtype: object
[23 49 19 74 85 58 81 84 69 77 30 72 88 18 38 89 42 87 73 50 61 32 70 59
 36 51 82 78 47 66 24 45 76 20 46 34 75 41 35 27 65 80 55 21 37 43 53 57
 29 25 44 64 67 48 63 62 31 68 71 54 79 39 40 60 56 28 26 83 33 52 86 22]


In [7]:
print(df.describe())   # Stats for numeric columns

               Age  BloodPressure
count  1995.000000    1782.000000
mean     53.249123     135.888328
std      20.507322      30.173937
min      18.000000      30.000000
25%      36.000000     112.000000
50%      53.000000     135.000000
75%      71.000000     157.000000
max      89.000000     400.000000


Based on the summary statistics, the `BloodPressure` column **contains outliers**.

---

### 🩺 **What Is the Ideal Range for Blood Pressure?**

According to medical guidelines (e.g., WHO, American Heart Association), **normal adult blood pressure** is generally considered to be:

| Category             | Systolic (upper) | Diastolic (lower) |
| -------------------- | ---------------- | ----------------- |
| Normal               | **< 120**        | **< 80**          |
| Elevated             | **120–129**      | **< 80**          |
| Hypertension Stage 1 | **130–139**      | **80–89**         |
| Hypertension Stage 2 | **140+**         | **90+**           |
| Hypertensive Crisis  | **180+**         | **120+**          |

In practice, **a realistic/physiological range** for systolic blood pressure values is roughly:

> **🟢 Normal range: 90–180 mmHg**
> Values **below 80** or **above 200** are often suspect (either outliers or entry errors).

---

### 🚨 The Stats

| Stat     | Value                  |
| -------- | ---------------------- |
| **Mean** | 135.9                  |
| **Min**  | 30.0 ❌ extremely low   |
| **Max**  | 400.0 ❌ extremely high |

So we definitely have:

* **Unrealistically low** values: `< 60`
* **Unrealistically high** values: `> 200` or especially `> 250`

---

### ✅ What we Can Do

In [8]:
# 1. Detect and Flag Outliers

# Using an acceptable range (say 60 to 200):

# Define acceptable range
valid_range = (60, 200)

# Flag outliers
outliers = df[(df['BloodPressure'] < valid_range[0]) | (df['BloodPressure'] > valid_range[1])]
print(outliers)

         Name  Age  BloodPressure Diagnosis
56    charlie   55          400.0   OBESITY
454     alice   68          300.0   obesity
515     diana   82          400.0  Diabetes
643     FRANK   83          300.0    Asthma
899    Hannah   42          400.0   Obesity
1006    alice   81           30.0   HEALTHY
1054    Alice   26          400.0  Diabetes
1114  Charlie   45          400.0   HEALTHY
1489    FRANK   62          300.0  DIABETES
1785   HANNAH   77           30.0   OBESITY


#2. Handle Outliers

* **Option 1:** Remove them

```python
df = df[(df['BloodPressure'] >= 60) & (df['BloodPressure'] <= 200)]
```

* **Option 2:** Impute with median or cap values

```python
# Cap extreme values
df['BloodPressure'] = df['BloodPressure'].clip(lower=60, upper=200)
```

In [9]:
# Cap extreme values
df['BloodPressure'] = df['BloodPressure'].clip(lower=60, upper=200)

In [10]:
print(df.head())       # Preview first few rows

      Name  Age  BloodPressure     Diagnosis
0   Hannah   23          153.0  HYPERTENSION
1    FRANK   49          128.0        asthma
2    Diana   19          163.0       Healthy
3  charlie   74          135.0      Diabetes
4    FRANK   85           94.0       Obesity


#### 🧊 3. **Check for Missing Values**

```python
print(df.isnull().sum())
```

✅ **What to Look For**:

* Any column with non-zero missing count needs fixing

In [11]:
print(df.isnull().sum())

Name               0
Age                0
BloodPressure    213
Diagnosis          0
dtype: int64


We're looking at **213 missing values** in the `BloodPressure` column — that’s about **10.7%** of the 1995 total entries.

---

### 🛠️ **What Can We Do with Missing Values?**

We’ve got a few options:

---

### ✅ **Option 1: Impute (Fill in the Gaps)**

#### **1. Fill with the Median**

This is the most common, especially for numeric medical data:

```python
df['BloodPressure'].fillna(df['BloodPressure'].median(), inplace=True)
```

* ✔ Robust to outliers
* ✔ Keeps all rows

#### **2. Fill with Mean**

```python
df['BloodPressure'].fillna(df['BloodPressure'].mean(), inplace=True)
```

#### **3. Fill Using Age Groups or Diagnosis**

If blood pressure correlates with age or diagnosis, we can group by those:

```python
df['BloodPressure'] = df.groupby('Age')['BloodPressure'].transform(lambda x: x.fillna(x.median()))
```

or:

```python
df['BloodPressure'] = df.groupby('Diagnosis')['BloodPressure'].transform(lambda x: x.fillna(x.median()))
```

---

### 🧹 **Option 2: Drop the Missing Rows**

Only do this if the rows with missing BP aren't important:

```python
df.dropna(subset=['BloodPressure'], inplace=True)
```

* ❌ We lose 213 rows (over 10% of the data)
* ✔ Simple and safe if we're not analyzing or predicting BP

---

### ⚠️ Summary

| Strategy                        | When to Use                                            |
| ------------------------------- | ------------------------------------------------------ |
| `fillna(median)`                | Default for numeric columns with mild outliers         |
| Group-based fill (e.g., by Age) | When BP likely depends on another column               |
| Drop rows                       | When missing BP rows are irrelevant or small in number |

In [12]:
df['BloodPressure'] = df['BloodPressure'].fillna(df['BloodPressure'].median())

#### 📛 4. **Check for Duplicates**

```python
print(df.duplicated().sum())
```

✅ **What to Look For**:

* If >0 → You likely have repeated rows you should remove


In [13]:
print(df.duplicated().sum())

12


Getting `12` as the output means:

> ✅ **There are 12 completely duplicate rows in your DataFrame.**

### ✅ What We Can Do Next

#### 1. **View the Duplicates**

```python
print(df[df.duplicated()])
```

#### 2. **Drop the Duplicates**

```python
df = df.drop_duplicates()
```

#### 3. **Confirm They’re Gone**

```python
print(df.duplicated().sum())  # should now return 0
```

In [20]:
# 1. View the Duplicates

print(df[df.duplicated()])

# 2. Drop the Duplicates

df = df.drop_duplicates()

# 3. Confirm They’re Gone

print(df.duplicated().sum())  # should now return 0

        Name  Age  BloodPressure     Diagnosis
272    DIANA   18          129.0  hypertension
540    ALICE   42          135.0  Hypertension
542   Hannah   36          174.0       obesity
554      eve   85          136.0       Obesity
584    Frank   55          161.0       obesity
763    alice   32          150.0       Healthy
766    Grace   47          170.0       Obesity
915    frank   74          164.0        asthma
969      Ivy   44          135.0  Hypertension
1166     eve   21          124.0      Diabetes
1191    jack   76          165.0  HYPERTENSION
1285     EVE   55           98.0       obesity
0


#### 🔡 5. **Check for Inconsistent Formatting / Casing / Whitespace**

```python
print(df['Name'].unique())
```

✅ **What to Look For**:

* Same name spelled differently (e.g., "Bob", "bob ", "BOB")
* Extra spaces (`'Jack '`, `' jack'`)


In [14]:
print(df['Name'].unique())

['Hannah' 'FRANK' 'Diana' 'charlie' 'Charlie' 'Bob' 'EVE' 'alice' 'eve'
 'ALICE' 'grace' 'ivy' 'Alice' 'HANNAH' 'Jack' 'diana' 'Grace' 'frank'
 'Eve' 'Frank' 'Ivy' 'GRACE' 'hannah' 'JACK' 'bob' 'DIANA' 'frannk' 'BOB'
 'CHARLIE' 'IVY' 'jack' 'Iyv' 'AALICE' 'BBOB' 'JAKC' 'ffrank' 'ccharlie'
 'farnk' 'DDiana' 'RFANK' 'AJCK' 'aliice' 'diiana' 'Hannnah' 'Chharlie'
 'frnak' 'frrank' 'FRAKN' 'allice' 'ALCIE' 'Jacck' 'hnanah' 'FRAANK'
 'jakc' 'CCHARLIE' 'AAlice' 'eeve' 'IIVY' 'rfank']


## ✅ What’s Wrong

From the output, we can see:

### 1. **Casing Issues**

* `'Bob'`, `'BOB'`, `'bob'`
* `'Alice'`, `'ALICE'`, `'alice'`, etc.

### 2. **Misspellings / Variants**

* `'frannk'`, `'frrank'`, `'frnak'`, `'FRAKN'`, `'ffrank'` → probably all meant to be `'Frank'`
* `'diiana'`, `'DDiana'`, `'DIANA'` → likely `'Diana'`
* `'Jack'`, `'jack'`, `'JAKC'`, `'AJCK'`, `'Jacck'` → probably `'Jack'`
* `'charlie'`, `'ccharlie'`, `'CCHARLIE'`, `'Chharlie'` → likely `'Charlie'`

### 3. **Typos / Extra Characters**

* `'AALICE'`, `'aliice'`, `'allice'`, `'AAlice'`, `'ALCIE'` → probably `'Alice'`
* `'Iyv'`, `'IIVY'` → probably `'Ivy'`

---

## 🔧 How to Clean It


In [21]:
# Step 1: Strip Whitespace and Standardize Case

df['Name'] = df['Name'].str.strip().str.title()

# Step 2: Manually Correct Misspellings

name_corrections = {
    'Frrank': 'Frank',
    'Frnak': 'Frank',
    'Frannk': 'Frank',
    'Ffrrank': 'Frank',
    'Fraank': 'Frank',
    'Frakn': 'Frank',
    'Rfank': 'Frank',
    'Ffrank': 'Frank',

    'Diiana': 'Diana',
    'Ddiana': 'Diana',
    'Diana ': 'Diana',

    'Chharlie': 'Charlie',
    'Ccharlie': 'Charlie',
    'Ccharlie ': 'Charlie',

    'Jakc': 'Jack',
    'Jakc': 'Jack',
    'Ajck': 'Jack',
    'Jacck': 'Jack',

    'Aliice': 'Alice',
    'Allice': 'Alice',
    'Aalice': 'Alice',
    'Alcie': 'Alice',
    'Aaice': 'Alice',

    'Iyv': 'Ivy',
    'Iivy': 'Ivy',
    'Iiivy': 'Ivy',
}

df['Name'] = df['Name'].replace(name_corrections)

# Step 3: Recheck Uniques
print(sorted(df['Name'].unique()))

['Alice', 'Bbob', 'Bob', 'Charlie', 'Diana', 'Eeve', 'Eve', 'Farnk', 'Frank', 'Grace', 'Hannah', 'Hannnah', 'Hnanah', 'Ivy', 'Jack']


The list is now shorter, but there are still some **stragglers and misspellings** that should be cleaned up.

We will go through them:

---

### ❌ Remaining Misspellings / Inconsistencies:

| Incorrect | Intended |
| --------- | -------- |
| `Bbob`    | `Bob`    |
| `Eeve`    | `Eve`    |
| `Farnk`   | `Frank`  |
| `Hannnah` | `Hannah` |
| `Hnanah`  | `Hannah` |

---

### ✅ Final Fix

We can use this small mapping dictionary to correct them:


In [22]:
name_fixes = {
    'Bbob': 'Bob',
    'Eeve': 'Eve',
    'Farnk': 'Frank',
    'Hannnah': 'Hannah',
    'Hnanah': 'Hannah'
}

df['Name'] = df['Name'].replace(name_fixes)

# Then recheck:

print(sorted(df['Name'].unique()))

['Alice', 'Bob', 'Charlie', 'Diana', 'Eve', 'Frank', 'Grace', 'Hannah', 'Ivy', 'Jack']


📌 Use `.value_counts()` to inspect frequency:

```python
print(df['Diagnosis'].value_counts(dropna=False))
```

In [15]:
print(df['Diagnosis'].value_counts(dropna=False))

Diagnosis
DIABETES        166
Obesity         150
diabetes        149
Diabetes        147
Asthma          135
HEALTHY         133
OBESITY         131
healthy         130
Healthy         129
hypertension    128
ASTHMA          124
obesity         122
HYPERTENSION    119
Hypertension    119
asthma          113
Name: count, dtype: int64


The output makes it very clear: the `'Diagnosis'` column suffers from **inconsistent casing and formatting**, just like `'Name'` earlier.

---

### 🔍 Issues Identified:

We have the same diagnoses written in many different ways:

| Diagnosis      | Variants Found                                       |
| -------------- | ---------------------------------------------------- |
| `Diabetes`     | `'diabetes'`, `'Diabetes'`, `'DIABETES'`             |
| `Obesity`      | `'Obesity'`, `'OBESITY'`, `'obesity'`                |
| `Asthma`       | `'Asthma'`, `'ASTHMA'`, `'asthma'`                   |
| `Hypertension` | `'Hypertension'`, `'HYPERTENSION'`, `'hypertension'` |
| `Healthy`      | `'Healthy'`, `'HEALTHY'`, `'healthy'`                |

### ✅ Step-by-Step Fix

#### 1. Standardize Casing

```python
df['Diagnosis'] = df['Diagnosis'].str.strip().str.title()
```

This will convert all variants to a consistent format:

* `'DIABETES'`, `'diabetes'` → `'Diabetes'`
* `'HEALTHY'`, `'healthy'` → `'Healthy'`

---

#### 2. Recheck the Counts

```python
print(df['Diagnosis'].value_counts())
```

In [23]:
# 1. Standardize Casing

df['Diagnosis'] = df['Diagnosis'].str.strip().str.title()

# 2. Recheck the Counts

print(df['Diagnosis'].value_counts())

Diagnosis
Diabetes        461
Obesity         398
Healthy         391
Asthma          371
Hypertension    362
Name: count, dtype: int64


#### 🔠 6. **Check for Typos or Unexpected Categories**

You may find unusual or misspelled values in categorical columns.

```python
print(sorted(df['Diagnosis'].unique()))
```

✅ **What to Look For**:

* Close but different strings: `'Healthy'`, `'healthy'`, `'healty'`
* Capitalization issues

In [24]:
print(sorted(df['Diagnosis'].unique()))

['Asthma', 'Diabetes', 'Healthy', 'Hypertension', 'Obesity']


#### 🧮 7. **Check for Outliers in Numeric Columns**

Use `.describe()` and visual tools:

```python
print(df['Age'].describe())
print(df['BloodPressure'].describe())
```

✅ **What to Look For**:

* Extreme min or max values that don’t make sense
* For example: Age of 180, BP of 250


In [25]:
print(df['Age'].describe())

count    1983.000000
mean       53.276349
std        20.504861
min        18.000000
25%        36.000000
50%        53.000000
75%        71.000000
max        89.000000
Name: Age, dtype: float64


In [26]:
print(df['BloodPressure'].describe())


count    1983.000000
mean      135.111952
std        24.669430
min        60.000000
25%       115.000000
50%       135.000000
75%       154.000000
max       200.000000
Name: BloodPressure, dtype: float64


#### 📏 8. **Check for Incorrect Data Types**

```python
print(df.dtypes)
```

✅ **What to Look For**:

* `Age` or `BloodPressure` showing as `object` → May have non-numeric values
* Fix with: `df['Age'] = pd.to_numeric(df['Age'], errors='coerce')`

In [27]:
print(df.dtypes)

Name              object
Age                int64
BloodPressure    float64
Diagnosis         object
dtype: object


## 🧠 Summary Table

| Problem Type             | What to Check                  | Tools & Commands                      |
| ------------------------ | ------------------------------ | ------------------------------------- |
| Missing values           | `.isnull().sum()`              | Check for incomplete rows             |
| Duplicates               | `.duplicated().sum()`          | Identify repeated rows                |
| Typos / casing           | `.unique()`, `.value_counts()` | Spot inconsistent entries             |
| Outliers                 | `.describe()`, `boxplot()`     | Look for unusually high/low numbers   |
| Extra whitespace         | `.str.strip()`, `.unique()`    | Values with trailing/leading spaces   |
| Wrong data types         | `.dtypes`                      | Ensure numerics are not objects       |
| Category inconsistencies | `.value_counts()`              | Identify similar-but-different labels |


### 🛠 Pro Tip for Beginners:

Always begin with:

```python
df.shape
df.head()
df.info()
df.describe(include='all')
```

These 4 lines reveal **most cleaning issues** upfront.



### ✅ 7. **Save Cleaned Dataset**

```python
df.to_csv("patients_cleaned.csv", index=False)
```

In [28]:
# 7. Save Cleaned Dataset
df.to_csv("../../datasets/patients_cleaned.csv", index=False)

In [2]:
import pandas as pd
import numpy as np
import random
from datetime import datetime, timedelta

# Reproducibility
random.seed(42)
np.random.seed(42)

# Data pools
product_clean = ['Widget A', 'Widget B', 'Widget C', 'Widget D']
product_typos = ['widget a', 'Widget A ', 'Widgget B', 'widget C']
region_clean = ['North', 'South', 'East', 'West']
region_typos = ['north ', 'SOUTH', 'eest', 'West ']

# Dates
start_date = datetime.strptime('2023-01-01', '%Y-%m-%d')

# Number of rows
n_rows = 1900  # We'll add duplicates to reach 2000

# Create base data
dates = [start_date + timedelta(days=random.randint(0, 60)) for _ in range(n_rows)]
products = np.random.choice(product_clean + product_typos, size=n_rows)
regions = np.random.choice(region_clean + region_typos, size=n_rows)
sales = np.random.randint(500, 5000, size=n_rows).astype(float)

# Insert some missing values
missing_indices = np.random.choice(n_rows, size=100, replace=False)
for i in missing_indices:
    sales[i] = np.nan

# Insert some outliers
outlier_indices = np.random.choice(n_rows, size=20, replace=False)
for i in outlier_indices:
    sales[i] = np.random.randint(10000, 30000)

# Create DataFrame
df = pd.DataFrame({
    'Date': [d.strftime('%Y-%m-%d') for d in dates],
    'Product': products,
    'Region': regions,
    'Sales': sales
})

# Add duplicates
duplicates = df.sample(frac=0.05, random_state=42)
df = pd.concat([df, duplicates], ignore_index=True)

# Shuffle the data
df = df.sample(frac=1, random_state=42).reset_index(drop=True)

# Save to CSV
df.to_csv("unclean_sales.csv", index=False)
